File: NiceReads/Backend/TOS/Common/TOS.CQRS/Handlers/HandlerExecutor.cs
using System;
using System.Threading.Tasks;
using TOS.CQRS.Executions;
using TOS.CQRS.Logging;

namespace TOS.CQRS.Handlers
{
    public class HandlerExecutor : IHandlerExecutor
    {
        private readonly IHandlerExecutorLogger _handlerExecutorLogger;

        public HandlerExecutor(IHandlerExecutorLogger handlerExecutorLogger)
        {
            _handlerExecutorLogger = handlerExecutorLogger;
        }

        public void Execute<TExecution, THandler>(THandler executionHandler, TExecution execution)
            where TExecution : IExecutionRequest
            where THandler : IExecutionHandler<TExecution>
        {
            using (IHandlerExecutorLoggerScope logger = _handlerExecutorLogger.CreateScope(execution, executionHandler))
            {
                logger.BeforeExecution();
                try
                {
                    using (logger.TimeExecution())
                    {
                        executionHandler.Execute(execution);
                    }
                    logger.AfterExecution();
                }
                catch (Exception ex)
                {
                    logger.OnError(ex);
                    throw;
                }
            }
        }

        public TResult Execute<TExecution, THandler, TResult>(THandler executionHandler, TExecution execution)
            where TExecution : IExecutionRequest<TResult>
            where THandler : IExecutionHandler<TExecution, TResult>
        {
            using (IHandlerExecutorLoggerScope<TResult> logger = _handlerExecutorLogger.CreateScope<TExecution, THandler, TResult>(execution, executionHandler))
            {
                logger.BeforeExecution();
                try
                {
                    TResult result;
                    using (logger.TimeExecution())
                    {
                        result = executionHandler.Execute(execution);
                    }
                    logger.AfterExecution(result);
                    return result;
                }
                catch (Exception ex)
                {
                    logger.OnError(ex);
                    throw;
                }
            }
        }

        public async Task ExecuteAsync<TExecution, THandler>(THandler executionHandler, TExecution execution)
            where TExecution : IAsyncExecutionRequest
            where THandler : IAsyncExecutionHandler<TExecution>
        {
            using (IHandlerExecutorLoggerScope logger = _handlerExecutorLogger.CreateScopeForAsync(execution, executionHandler))
            {
                logger.BeforeExecution();
                try
                {
                    using (logger.TimeExecution())
                    {
                        await executionHandler.ExecuteAsync(execution);
                    }
                    logger.AfterExecution();
                }
                catch (Exception ex)
                {
                    logger.OnError(ex);
                    throw;
                }
            }
        }

        public async Task<TResult> ExecuteAsync<TExecution, THandler, TResult>(THandler executionHandler, TExecution execution)
            where TExecution : IAsyncExecutionRequest<TResult>
            where THandler : IAsyncExecutionHandler<TExecution, TResult>
        {
            using (IHandlerExecutorLoggerScope<TResult> logger = _handlerExecutorLogger.CreateScopeForAsync<TExecution, THandler, TResult>(execution, executionHandler))
            {
                logger.BeforeExecution();
                try
                {
                    TResult result;
                    using (logger.TimeExecution())
                    {
                        result = await executionHandler.ExecuteAsync(execution);
                    }
                    logger.AfterExecution(result);
                    return result;
                }
                catch (Exception ex)
                {
                    logger.OnError(ex);
                    throw;
                }
            }
        }
    }
}

File: NiceReads/Backend/Server/tos-server/src/cqrs/handlers/HandlerExecutor.ts
import { inject, injectable } from 'inversify';
import { Types } from '../../types';
import { IAsyncExecutionRequest, IAsyncExecutionRequestWithResult, IExecutionRequest, IExecutionRequestWitResult } from '../executions/execution.request';
import { IExecutionTimer } from '../logging/IExecutionTimer';
import { IHandlerExecutorLogger } from '../logging/IHandlerExecutorLogger';
import { IHandlerExecutorLoggerScope } from '../logging/IHandlerExecutorLoggerScope';
import { IHandlerExecutorLoggerWithResultScope } from '../logging/IHandlerExecutorLoggerWithResultScope';
import { IAsyncExecutionHandler } from './IAsyncExecutionHandler';
import { IAsyncExecutionHandlerWithResult } from './IAsyncExecutionHandlerWithResult';
import { IExecutionHandler } from './IExecutionHandler';
import { IExecutionHandlerWithResult } from './IExecutionHandlerWithResult';
import { IHandlerExecutor } from './IHandlerExecutor';

@injectable()
export class HandlerExecutor implements IHandlerExecutor {

    constructor(@inject(Types.HANDLER_EXECUTOR_LOGGER) private readonly handlerExecutorLogger: IHandlerExecutorLogger) { }

    public execute<TExecution extends IExecutionRequest, THandler extends IExecutionHandler<TExecution>>(
        executionHandler: THandler,
        execution: TExecution): void {

        const loggerScope: IHandlerExecutorLoggerScope = this.handlerExecutorLogger.createScope(execution, executionHandler);
        loggerScope.beforeExecution();
        try {
            const executionTimer: IExecutionTimer = loggerScope.timeExecution();
            try {
                executionHandler.execute(execution);
            } finally {
                executionTimer.stop();
            }
            loggerScope.afterExecution();
        } catch (error) {
            loggerScope.onError(error as Error);
            throw error;
        } finally {
            loggerScope.complete();
        }
    }

    public executeWithResult<
        TExecution extends IExecutionRequestWitResult<TResult>,
        THandler extends IExecutionHandlerWithResult<TExecution, TResult>, TResult>
    (executionHandler: THandler, execution: TExecution): TResult {

        const loggerScope: IHandlerExecutorLoggerWithResultScope<TResult> = this.handlerExecutorLogger
            .createScopeWithResult<TExecution, THandler, TResult>(execution, executionHandler);
        loggerScope.beforeExecution();
        try {
            let result: TResult;
            const executionTimer: IExecutionTimer = loggerScope.timeExecution();
            try {
                result = executionHandler.execute(execution);
            } finally {
                executionTimer.stop();
            }
            loggerScope.afterExecutionWithResult(result);
            return result;
        } catch (error) {
            loggerScope.onError(error as Error);
            throw error;
        } finally {
            loggerScope.complete();
        }
    }

    public async executeAsync<TExecution extends IAsyncExecutionRequest, THandler extends IAsyncExecutionHandler<TExecution>>(
        executionHandler: THandler,
        execution: TExecution): Promise<void> {

        const loggerScope: IHandlerExecutorLoggerScope = this.handlerExecutorLogger.createScopeForAsync(execution, executionHandler);
        loggerScope.beforeExecution();
        try {
            const executionTimer: IExecutionTimer = loggerScope.timeExecution();
            try {
                await executionHandler.executeAsync(execution);
            } finally {
                executionTimer.stop();
            }
            loggerScope.afterExecution();
        } catch (error) {
            loggerScope.onError(error as Error);
            throw error;
        } finally {
            loggerScope.complete();
        }
    }

    public async executeWithResultAsync<
        TExecution extends IAsyncExecutionRequestWithResult<TResult>,
        THandler extends IAsyncExecutionHandlerWithResult<TExecution, TResult>, TResult>
    (executionHandler: THandler, execution: TExecution): Promise<TResult> {

        const loggerScope: IHandlerExecutorLoggerWithResultScope<TResult> = this.handlerExecutorLogger
            .createScopeWithResultForAsync<TExecution, THandler, TResult>(execution, executionHandler);
        loggerScope.beforeExecution();
        try {
            let result: TResult;
            const executionTimer: IExecutionTimer = loggerScope.timeExecution();
            try {
                result = await executionHandler.executeAsync(execution);
            } finally {
                executionTimer.stop();
            }
            loggerScope.afterExecutionWithResult(result);
            return result;
        } catch (error) {
            loggerScope.onError(error as Error);
            throw error;
        } finally {
            loggerScope.complete();
        }
    }
}